covid_graphs/covid_graphs/formula.py

```diff
@@ -2,42 +2,27 @@
 import math
 from abc import abstractmethod
 from dataclasses import dataclass
-from typing import Callable
+from typing import Callable, List, Tuple
 
 import numpy as np
 
 from . import fit_atg_model
 from .country_report import CountryReport
 
 
+@dataclass
 class Curve:
-    def __init__(
-        self,
-        func: Callable[[datetime.date], float],
-        start_date: datetime.date,
-        end_date: datetime.date,
-        label: str,
-    ) -> None:
-        """
-        Calculates function trace for dates in a half-open interval [start_date, end_date).
-
-        Sets the trace as properties `xs` and `ys`. Also calculates a point (x_max, y_max)
-        for which the maximum is achieved.
-
-        Note: The current design assumes the range of the plot is known at initialisation.
-        """
-        self.start_date = start_date
-        self.end_date = end_date
-        self.label = label
-
-        self.xs = [
-            start_date + datetime.timedelta(days=d) for d in range((end_date - start_date).days)
-        ]
-        self.ys = [func(x) for x in self.xs]
-
-        idx_max = np.argmax(self.ys)
-        self.x_max = self.xs[idx_max]
-        self.y_max = self.ys[idx_max]
+    func: Callable
+    start_date: datetime.date
+    display_at_least_until: datetime.date
+    label: str
+
+    def generate_trace(self, end_date: datetime.date) -> Tuple[List[datetime.date], np.ndarray]:
+        """Generates trace corresponding to the half-open interval [self.start_date, end_date)"""
+        raw_xs = np.arange((end_date - self.start_date).days)
```

Comment:
Oh, you changed `func` to be `float -> float`. I think this is a step back in the design, in my opinion the logic of how to convert `date` to `x` values should be the responsibility of Formula generator, and should not leak here. Is there a reason for this change?

Reply:
I thought about it a bit more. The real question here is who should be aware of how to convert date to float.

Reply:
The previous API was convoluted and prone to errors. Even the author himself made some off-by-one errors that I've fixed :smile: 

We had to calculate `(date - day_zero).days` at three different places for no good reason. Now it's simpler:

- We define a function, e.g. `x^2 / e^(x/5)`
- We say where `x=0` is (i.e. set `start_date`).
- Profit! Fewer bugs and everyone is happier.

There's no reason to convert dates to function values at many places, the only place this is done is in `generate_trace`. Much cleaner IMO!